{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a57218e2-3704-4696-975d-551f4569d025",
   "metadata": {},
   "source": [
    "### Challenge 05.\n",
    "### Translate DNA Sequences to Proteins.\n",
    "### *Taken from [Geeks for Geeks.](https://www.geeksforgeeks.org/dna-protein-python-3/)*\n",
    "\n",
    "**Description**\n",
    "\n",
    "*Translation Theory :* DNA ⇒ RNA ⇒ Protein\n",
    "\n",
    "Life depends on the ability of cells to store, retrieve, and translate genetic instructions. These instructions are needed to make and maintain living organisms. For a long time, it was not clear what molecules were able to copy and transmit genetic information. We now know that this information is carried by the deoxyribonucleic acid or DNA in all living things. \n",
    "\n",
    "DNA: DNA is a discrete code physically present in almost every cell of an organism. We can think of DNA as a one dimensional string of characters with four characters to choose from. These characters are A, C, G, and T. They stand for the first letters with the four nucleotides used to construct DNA. The full names of these nucleotides are Adenine, Cytosine, Guanine, and Thymine. Each unique three character sequence of nucleotides, sometimes called a nucleotide triplet, corresponds to one amino acid. The sequence of amino acids is unique for each type of protein and all proteins are built from the same set of just 20 amino acids for all living things. \n",
    "\n",
    "Instructions in the DNA are first transcribed into RNA and the RNA is then translated into proteins. We can think of DNA, when read as sequences of three letters, as a dictionary of life. \n",
    "\n",
    "*Aim:* Convert a given sequence of DNA into its Protein equivalent. \n",
    "\n",
    "*Source:* Download a DNA strand as a text file from a public web-based repository of DNA sequences from NCBI. The Nucleotide sample is (NM_207618.2), which can be found [here](https://www.ncbi.nlm.nih.gov/nuccore/NM_207618.2)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e1c9acc-5f73-42ed-a20e-9d01d77c39b0",
   "metadata": {},
   "source": [
    "### Proposed Solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "04ce074a-ed34-458f-b102-a9cff676e2a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Challenge 05\n",
    "### Translate DNA Sequences to Proteins\n",
    "### Proposed Solution\n",
    "\n",
    "\n",
    "### Import libraries\n",
    "import json\n",
    "\n",
    "\n",
    "\n",
    "### Load utilities to translate\n",
    "\n",
    "# Load list of aminoacids\n",
    "def load_aminoacids():\n",
    "\n",
    "    # Initialize list of aminoacids\n",
    "    ls_aminoacids = []\n",
    "\n",
    "    # Fill list of aminoacids\n",
    "    with open('input_aminoacids.dat') as file_aminoacids:\n",
    "        for line in file_aminoacids:\n",
    "            # Whitespaces are deleted, characters to uppercase, and splitting information into columns\n",
    "            aminoacid_info = line.replace(\" \", \"\").replace(\"\\n\", \"\").upper().split(\"-\")\n",
    "            # Append extracted aminoacid information\n",
    "            ls_aminoacids.append(aminoacid_info)\n",
    "\n",
    "    return ls_aminoacids\n",
    "\n",
    "\n",
    "# Load dictionary of codons (3-nucleotids-in-a-row) to aminoacids\n",
    "def load_codons(ls_aminoacids):\n",
    "    \n",
    "    # Opening database of codons\n",
    "    file_codons = open('input_codon_to_aminoacids.json')\n",
    "    # Convert JSON object to a dictionary\n",
    "    dc_codons = json.load(file_codons)\n",
    "    # List of aminoacids - names of 3 characters\n",
    "    ls_aminoacids_abbrv_3char = [aminoacid[1] for aminoacid in ls_aminoacids]+['STOP']\n",
    "\n",
    "    # Assign new values to dictionary, 1 character per aminoacid instead of 3 characters\n",
    "    for key, value in dc_codons.items():\n",
    "        # Convert the value to uppercase\n",
    "        value_aux = value.upper()\n",
    "        # Special case for stop codons\n",
    "        if value_aux == \"STOP\":\n",
    "            dc_codons[key] = \"_\"\n",
    "        # Normal cases\n",
    "        else:\n",
    "            # Index of the aminoacid\n",
    "            index_aux = ls_aminoacids_abbrv_3char.index(value_aux)\n",
    "            # Equivalent name of 1 character\n",
    "            aminoacid_1char = ls_aminoacids[index_aux][2]\n",
    "            # New value in the dictionary for the codon\n",
    "            dc_codons[key] = aminoacid_1char\n",
    "        \n",
    "    return dc_codons\n",
    "\n",
    "\n",
    "\n",
    "### Generate translation \n",
    "\n",
    "# Load input (DNA sequence)\n",
    "def load_input():\n",
    "    with open('input_dna_sequence.dat') as file_dna_sequence:\n",
    "        # The entire text processing is wrapped in one line\n",
    "        dna_chain = \"\".join([line.replace(\"\\n\", \"\").replace(\" \", \"\").upper() for line in file_dna_sequence])\n",
    "    return dna_chain\n",
    "\n",
    "\n",
    "# RNA transcription\n",
    "def rna_chain(dna_chain):\n",
    "    # Just replacing nucleotide \"T\" by \"U\"\n",
    "    dna_chain = dna_chain.upper().replace(\"T\", \"U\")\n",
    "    return dna_chain\n",
    "\n",
    "\n",
    "# Find start and end of the RNA sequence\n",
    "def crop_rna_sequence(rna):\n",
    "\n",
    "    # Look for the 3-characters sequence \"AUG\" as they set the beginning of a protein\n",
    "    while rna[0:3] != \"AUG\":\n",
    "        rna = rna[1:]\n",
    "\n",
    "    # Look for the 3-characters sequences that set the end of a protein\n",
    "    # Initial guess\n",
    "    final_index = 3\n",
    "    possible_end = rna[final_index:final_index+3]\n",
    "    # Iterative search from the starting aminoacid \"AUG\"\n",
    "    while (possible_end != \"UAA\") and (possible_end != \"UAG\") and (possible_end != \"UGA\"):\n",
    "        final_index += 3\n",
    "        possible_end = rna[final_index:final_index+3]\n",
    "    rna = rna[0:final_index]\n",
    "\n",
    "    return rna\n",
    "\n",
    "\n",
    "# Convert RNA into aminoacids\n",
    "# Complete translation for a single protein\n",
    "def rna2protein(rna, dc_codons):\n",
    "    \n",
    "    # Lenght of the RNA chain\n",
    "    lrna = len(rna)\n",
    "\n",
    "    # Empty string where sequence of aminoacids will be saved\n",
    "    sequence_aminoacids = \"\"\n",
    "    \n",
    "    # Translation of RNA into aminoacids\n",
    "    for ii in range(int(lrna/3)):\n",
    "        # Extract codon\n",
    "        codon = rna[3*ii:3*(ii+1)]\n",
    "        # Add aminoacid to the sequence\n",
    "        aminoacid_1char = dc_codons[codon]\n",
    "        sequence_aminoacids = sequence_aminoacids + aminoacid_1char\n",
    "\n",
    "    return sequence_aminoacids\n",
    "\n",
    "\n",
    "# Verify result\n",
    "def validation(real_res):\n",
    "    # Load expected result\n",
    "    with open('output_protein_sequence.dat') as file_protein_sequence:\n",
    "        exp_res = \"\".join([line.replace(\"\\n\", \"\").replace(\" \", \"\").upper() for line in file_protein_sequence])\n",
    "    # Verification\n",
    "    check_val = real_res == exp_res\n",
    "    return check_val\n",
    "\n",
    "\n",
    "\n",
    "### Full sequence of aminoacids for a single protein (main function)\n",
    "def main():\n",
    "\n",
    "    # Load list of 20 fundamental aminoacids\n",
    "    ls_aminoacids = load_aminoacids()\n",
    "\n",
    "    # Load dictionary of codons\n",
    "    # Each combination between nucleotides \"U\", \"C\", \"A\", and \"G\" represent an aminoacid \n",
    "    dc_codons = load_codons(ls_aminoacids)\n",
    "\n",
    "    # Load DNA sequence\n",
    "    dna_chain = load_input()\n",
    "\n",
    "    # Transcription of DNA sequence into RNA\n",
    "    rna = rna_chain(dna_chain)\n",
    "\n",
    "    # Crop RNA from the starting to end aminoacids\n",
    "    rna = crop_rna_sequence(rna)\n",
    "\n",
    "    # RNA translated into sequence of aminoacids\n",
    "    sequence_aminoacids = rna2protein(rna, dc_codons)\n",
    "\n",
    "    # Perform result validation\n",
    "    check_val = validation(sequence_aminoacids)\n",
    "\n",
    "    return dna_chain, sequence_aminoacids, check_val"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "825baf27-21a9-4264-bee4-01ac0aef2586",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--------------------- Is it the expected result? ---------------------\n",
      "True\n",
      "--------------------- DNA sequence ---------------------\n",
      "GGTCAGAAAAAGCCCTCTCCATGTCTACTCACGATACATCCCTGAAAACCACTGAGGAAGTGGCTTTTCAGATCATCTTGCTTTGCCAGTTTGGGGTTGGGACTTTTGCCAATGTATTTCTCTTTGTCTATAATTTCTCTCCAATCTCGACTGGTTCTAAACAGAGGCCCAGACAAGTGATTTTAAGACACATGGCTGTGGCCAATGCCTTAACTCTCTTCCTCACTATATTTCCAAACAACATGATGACTTTTGCTCCAATTATTCCTCAAACTGACCTCAAATGTAAATTAGAATTCTTCACTCGCCTCGTGGCAAGAAGCACAAACTTGTGTTCAACTTGTGTTCTGAGTATCCATCAGTTTGTCACACTTGTTCCTGTTAATTCAGGTAAAGGAATACTCAGAGCAAGTGTCACAAACATGGCAAGTTATTCTTGTTACAGTTGTTGGTTCTTCAGTGTCTTAAATAACATCTACATTCCAATTAAGGTCACTGGTCCACAGTTAACAGACAATAACAATAACTCTAAAAGCAAGTTGTTCTGTTCCACTTCTGATTTCAGTGTAGGCATTGTCTTCTTGAGGTTTGCCCATGATGCCACATTCATGAGCATCATGGTCTGGACCAGTGTCTCCATGGTACTTCTCCTCCATAGACATTGTCAGAGAATGCAGTACATATTCACTCTCAATCAGGACCCCAGGGGCCAAGCAGAGACCACAGCAACCCATACTATCCTGATGCTGGTAGTCACATTTGTTGGCTTTTATCTTCTAAGTCTTATTTGTATCATCTTTTACACCTATTTTATATATTCTCATCATTCCCTGAGGCATTGCAATGACATTTTGGTTTCGGGTTTCCCTACAATTTCTCCTTTACTGTTGACCTTCAGAGACCCTAAGGGTCCTTGTTCTGTGTTCTTCAACTGTTGAAAGCCAGAGTCACTAAAAATGCCAAACACAGAAGACAGCTTTGCTAATACCATTAAATACTTTATTCCATAAATATGTTTTTAAAAGCTTGTATGAACAAGGTATGGTGCTCACTGCTATACTTATAAAAGAGTAAGGTTATAATCACTTGTTGATATGAAAAGATTTCTGGTTGGAATCTGATTGAAACAGTGAGTTATTCACCACCCTCCATTCTCT\n",
      "--------------------- Translated sequence of aminoacids ---------------------\n",
      "MSTHDTSLKTTEEVAFQIILLCQFGVGTFANVFLFVYNFSPISTGSKQRPRQVILRHMAVANALTLFLTIFPNNMMTFAPIIPQTDLKCKLEFFTRLVARSTNLCSTCVLSIHQFVTLVPVNSGKGILRASVTNMASYSCYSCWFFSVLNNIYIPIKVTGPQLTDNNNNSKSKLFCSTSDFSVGIVFLRFAHDATFMSIMVWTSVSMVLLLHRHCQRMQYIFTLNQDPRGQAETTATHTILMLVVTFVGFYLLSLICIIFYTYFIYSHHSLRHCNDILVSGFPTISPLLLTFRDPKGPCSVFFNC\n"
     ]
    }
   ],
   "source": [
    "### Testing Proposed Solution\n",
    "### Translate a DNA sequence into an aminoacid sequence for a given protein (execute the main function)\n",
    "\n",
    "# Call the main function\n",
    "if __name__ == '__main__':\n",
    "    dna_chain, sequence_aminoacids, check_val = main()\n",
    "    print(\"--------------------- Is it the expected result? ---------------------\")\n",
    "    print(check_val)\n",
    "    print(\"--------------------- DNA sequence ---------------------\")\n",
    "    print(dna_chain)\n",
    "    print(\"--------------------- Translated sequence of aminoacids ---------------------\")\n",
    "    print(sequence_aminoacids)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "950588e5-0874-4f2b-acb6-609bff3c61cd",
   "metadata": {},
   "source": [
    "### [External Solution](https://www.geeksforgeeks.org/dna-protein-python-3/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "5407cb89-37a3-4780-af21-b4ce04b108d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Challenge 05\n",
    "### DNA to Protein Translation\n",
    "### External Solution\n",
    "\n",
    "\n",
    "\n",
    "### Definition of the solution\n",
    "\n",
    "# Function to read DNA chain\n",
    "def read_seq(inputfile):\n",
    "    with open(inputfile, \"r\") as f:\n",
    "        seq = f.read()\n",
    "    seq = seq.replace(\"\\n\", \"\")\n",
    "    seq = seq.replace(\"\\r\", \"\")\n",
    "    return seq\n",
    "\n",
    "# Function to translate DNA to protein\n",
    "def translate(seq):\n",
    "    \n",
    "    table = {\n",
    "        'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "        'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "        'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "        'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "        'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "        'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "        'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "        'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "        'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "        'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "        'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "        'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "        'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "        'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "        'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "        'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',\n",
    "    }\n",
    "    protein =\"\"\n",
    "    if len(seq)%3 == 0:\n",
    "        for i in range(0, len(seq), 3):\n",
    "            codon = seq[i:i + 3]\n",
    "            protein+= table[codon]\n",
    "    return protein\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "2bde0537-c448-4187-afe9-71cbf4a9a04f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein: MSTHDTSLKTTEEVAFQIILLCQFGVGTFANVFLFVYNFSPISTGSKQRPRQVILRHMAVANALTLFLTIFPNNMMTFAPIIPQTDLKCKLEFFTRLVARSTNLCSTCVLSIHQFVTLVPVNSGKGILRASVTNMASYSCYSCWFFSVLNNIYIPIKVTGPQLTDNNNNSKSKLFCSTSDFSVGIVFLRFAHDATFMSIMVWTSVSMVLLLHRHCQRMQYIFTLNQDPRGQAETTATHTILMLVVTFVGFYLLSLICIIFYTYFIYSHHSLRHCNDILVSGFPTISPLLLTFRDPKGPCSVFFNC\n"
     ]
    }
   ],
   "source": [
    "### Testing External Solution\n",
    "\n",
    "# Initialize inputs\n",
    "dna = read_seq(\"DNA_sequence_original.txt\")\n",
    "expected_protein = read_seq(\"amino_acid_sequence_original.txt\")\n",
    "\n",
    "# Execution\n",
    "protein = translate(dna[20:935])\n",
    "if protein == expected_protein:\n",
    "    print('Protein:', protein)\n",
    "else:\n",
    "    print('Bad DNA translation')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60848f68-4585-4dd8-835c-821e43a4cb70",
   "metadata": {},
   "source": [
    "### Final Remarks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f9add48-32e5-4284-8384-d154dcb27949",
   "metadata": {},
   "source": [
    "After challenge 04, I was fascinated about DNA. It is impressive how a single long chain of characters has the entire genetic information of a living being. I found this challenge in [Geeks for Geeks](https://www.geeksforgeeks.org/dna-protein-python-3/) about DNA, and I do not hesitate to face it. \n",
    "\n",
    "The external solution is the one given on the aforementioned website. The proposed solution is my own approach, where I even looked for the data sources by my own. This is the reason why both solutions use different inputs. The proposed solution requires as inputs the following files `input_aminoacids.dat`, `input_codon_to_aminoacids.json`, `input_dna_sequence.dat`, and `output_protein_sequence.dat`, while the external solution uses the inputs `DNA_sequence_original.txt` and `amino_acid_sequence_original.txt`. Thus, the proposed solution handles more data formats and consequently the data loading process is a little bit longer. Besides, the proposed solution was devised to automatically detect the start and the end of a protein. This is not achieved by the external solution.\n",
    "\n",
    "A more detail description of the DNA translation and the proposed solution can be found [here](https://github.com/wgfajardom/portfolio/blob/develop/Challenges/05_dna_to_protein_translation/explanation_dna_to_protein_translation.md)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "anaconda-panel-2023.05-py310",
   "language": "python",
   "name": "conda-env-anaconda-panel-2023.05-py310-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
